### DISCUSSION TITLE
Merge Queue Feedback: Support for semi-linear merge

### DISCUSSION THREAD

**Discussion by sandstrom** [community/community#14863] [Pull Requests]
We'd like to use the semi-linear merge strategy for our git repo (rebase on master before merge & keep merge commit). This is what we want our Git log to look like (below).

The benefit with this structure is that you can use any Git tool to list only merge commits[1] (show a summary of only the PRs), but still opt to see all individual commits where needed.
    
The merge queue should produce this same result in our master branch.
    
    ```
    *   c8be0c632 Merge pull request #1538 from my-org/api-error-logging
    |\  
    | * 80ecc4985 Fix security warning, bump nokogiri
    | * 750613638 Log and respond with more detailed validation errors in the API
    | * 0165d6812 Log code and details when rendering an API error response.
    | * 1d4daab48 Refactor email validation result to include a descriptive message
    | * 635214092 Move media_type logging into context_logging
    |/  
    *   1cccd4412 Merge pull request #1539 from my-org/profile-clarify
    |\  
    | * 87b342a32 Rename profile default to migration target
    | * 2515c1e59 Fix disallow removing last profile in company
    |/  
    *   b8f3f1658 Merge pull request #1540 from my-org/customer
    |\  
    | * 064b31232 Add customer-specific taxed allowance reduction
    |/  
    *   3cf449f94 Merge pull request #1528 from my-org/console-logging
    |\  
    | * 99657f212 Don't log to rails console in production
    |/  
    *   4d5c38544 Merge pull request #1535 from my-org/ingestion
    |\  
    | * cb001c129 Add customer to ingestion
    |/  
    *   8c72e7f19 Merge pull request #1527 from my-org/gemfile
    ```

[1] For example `git log --merges --decorate --oneline` and most Git UIs and e.g. VS Code Git UI has the same thing.

***

Another way of visualizing semi-linear merge is this illustration (semi-linear merge is the rightmost alternative).

![image](https://user-images.githubusercontent.com/122287/228825036-cf652698-1a6c-4602-82c8-7cfdc44cab88.png)

***

This functionality exists in Azure DevOps and Gitlab:

- https://devblogs.microsoft.com/devops/pull-requests-with-rebase/#semi-linear-merge
- https://docs.gitlab.com/user/project/merge_requests/methods/#merge-commit-with-semi-linear-history

**Comment by Gallimathias** [comment]
2. I would only partially agree with that. It can be quite handy when trying to understand what is happening on the queue. So for debugging reasons or also if you want to do something specific.

Maybe tying it to a "user right" would be a good solution to the problem.

**Reply by sandstrom** [reply]
Good point! 

Number 2 is the least painful of these, it's more of a nuisance, but I agree, making it configurable would be good.

**Comment by lukaspiatkowski** [comment]
AFAIK this shape of git history can be enforced in a non-queue branch with regular merge commits (not squashed, not rebased) and the "Require branches to be up to date before merging" option set to true. It would be quite nice if Merge Queue could help achieve the same (Rebase and Merge).

**Comment by at-ng** [comment]
Related discussion https://github.com/community/community/discussions/8940

**Comment by ghost** [comment]
> This is what we want our git log to look like (below):
> 
> The benefit with this structure is that you can use any Git tool to list only merge commits[1] (show a summary of only the PRs), but still opt to see all individual commits where needed.
>     
> The merge queue should produce this same result in our master branch.
>     
>     ```
>     *   c8be0c632 Merge pull request #1538 from my-org/api-error-logging
>     |\  
>     | * 80ecc4985 Fix security warning, bump nokogiri
>     | * 750613638 Log and respond with more detailed validation errors in the API
>     | * 0165d6812 Log code and details when rendering an API error response.
>     | * 1d4daab48 Refactor email validation result to include a descriptive message
>     | * 635214092 Move media_type logging into context_logging
>     |/  
>     *   1cccd4412 Merge pull request #1539 from my-org/profile-clarify
>     |\  
>     | * 87b342a32 Rename profile default to migration target
>     | * 2515c1e59 Fix disallow removing last profile in company
>     |/  
>     *   b8f3f1658 Merge pull request #1540 from my-org/customer
>     |\  
>     | * 064b31232 Add customer-specific taxed allowance reduction
>     |/  
>     *   3cf449f94 Merge pull request #1528 from my-org/console-logging
>     |\  
>     | * 99657f212 Don't log to rails console in production
>     |/  
>     *   4d5c38544 Merge pull request #1535 from my-org/ingestion
>     |\  
>     | * cb001c129 Add customer to ingestion
>     |/  
>     *   8c72e7f19 Merge pull request #1527 from my-org/gemfile
>     ```
> 
> 
> [1] For example `git log --merges --decorate --oneline` and most Git UIs and e.g. VS Code Git UI has the same thing.
> 
> *EDIT: Removed two points of feedback that had been fixed by Github.*

**Comment by kEND** [comment]
Adding our upvote for this.

We have a semi-linear commit history as described above and want to make use of the merge queue to replicate it.  

Rebase branch.  Merge to `main` using a merge commit.

**Reply by sandstrom** [reply]
@kEND Write to Github support and ask for it, that might help putting it on the radar.

Or ask all your colleagues to upvote this issue.

**Comment by sandstrom** [comment]
Gitlab has offered support for semi-linear merge for PRs for a while, but has also added support for it to "Merge Trains" (their Merge Queue) this week.

https://gitlab.com/gitlab-org/gitlab/-/commit/d353265111691fcabf2abe39403a94b01b6f1ecc

Anyone on the Github team who has any updates on this one?

It's the one thing blocking us from using the merge queue, and semi-linear merge is supported by other big SCMs (Azure DevOps, Gitlab, etc), so it's somewhat surprising that Github doesn't support it.

**Reply by torressam333** [reply]
Who can we tag here from GitHub to at least get some feedback on multiple merge methods for Merge Queue's?

@BillGates?

**Reply by sandstrom** [reply]
@torressam333 There is an entry about this that you can upvote on the original feedback-thread for the merge queue:

https://github.com/orgs/community/discussions/46757#discussioncomment-4907971

If you have many colleagues in your company that would also want this feature, ask all of them to upvote that issue (and upvote the top-level issue in this thread too). If you ask them in e.g. an internal chat tool at your company (or personally, even better) that should help surface this as an issue to prioritize for Github. I think upvotes is a fairly strong driver, though obviously not the only factor.

**Comment by elsnerpaul** [comment]
Stumbled over this, since we want exactly this feature. Is there in 2026 a way to configure this?

**Reply by sandstrom** [reply]
No. Send a message to Github support and ask for it.

**Comment by tylerharris82836-netizen** [comment]
A semi-linear merge queue would definitely make PR history much easier to scan while still preserving commit context. The balance between readability and detail feels as practical as [sales coaching for sdrs](https://lemcoach.com/for-sdrs), where structured progress matters without losing the full picture.
